### ISSUE TITLE
Current master is broken, set master to a known good point and focus efforts on moving fowards safely

### ISSUE THREAD

**Issue by hornc** [internetarchive/infogami#25]
Current master 84fa0aa4ad4a585c138b165685cf89da271208ff has introduced errors.

The current Open Library production instance is operating at 7595ae764b045c58c762d303da519e00cfe06078 

(unfortunately) this qualifies as our last and best "known good" point. Ideally we would like to have known good and regularly run tests to protect us, but it is not clear that the existing tests have been fully run in any environment recently (i.e. for years), and CI testing has only been added to this recently and have never actually passed in a meaningful way.

#23 addresses the testing issue

This is to set master to a known good point for production Open Library to pull from, and allow us to add additional commits in pursuit of better tests and compatibility in a more controlled fashion.

It is likely we will have to revert some recent changes to branches and move smaller changesets across, once CI testing has stabilised.

**Comment by cclauss** [comment]
#5 was largely created with tools (a targeted version of __futurize --stage1 -w .__) so if we want to revert #5 and then make sure CI is up to snuff than I can recreate #5 from scratch.  As you wish.

**Comment by mekarpeles** [comment]
@hornc as per earlier suggestions + @cdini, proposing the branch https://github.com/internetarchive/infogami/tree/python3-dev for further development

**Comment by cclauss** [comment]
@hornc's approach above makes more sense to me.

EMPHASIS ON... I am NOT a fan of a separate branch for Python 3.   I have seen this fail _many_ times.  Python porting best practice is a single codebase that supports both Python 2 and Python 3 so that A/B testing can go on the entire time.  We need to know immediately if Python 3 changes break Python 2 compatibility.  https://docs.python.org/3/howto/pyporting.html

**Comment by mekarpeles** [comment]
I'll leave it up to @hornc how he feels it's best to move forward (@cdini and I offered our feedback and have full confidence in whatever direction @hornc chooses). The general idea he described above works for me. Assuming our tests are repaired and in place, I have no problem with us using master, so long as a merge doesn't propagate to the production service (which pulls from this infogami)

**Comment by cdrini** [comment]
The branch wouldn't be to hold Python 3 stuff; it would just be a place to put all the commits that are _currently_ untested. We want master to always be production ready. But we have no way to guarantee production readiness right now. NO NEW COMMITS SHOULD GO INTO THE PYTHON BRANCH. Once we have some way to automatically check production readiness, _then_ we will merge in the commits from the branch into master. After _that_, we can start looking at open pull requests to merge in the rest. (Let's try to come up with a good name for that branch that communicates this).

I don't see any way we can move forward with python 2->3 until we have _some_ sort of automatic way to check that things aren't breaking. THIS SUCKS. But I can't see any way around it; the first step is setting up Travis/testing/etc to make sure we don't break things. This repo handles most of the database access; this is the code that drives all of Open Library; we _really_ can't afford to do the "let's push the code on production data and see what happens" approach.

@cclauss Does that make sense? Am I missing anything? Is there a way we could get to accepting pull requests faster?

**Comment by cclauss** [comment]
Travis CI was started on this repo in the final days of 2018 #4 

That plan _would_ make sense if the symlink between OL and infogami was not a stale one.  As it is, I do not see how changes in this repo effect OL at all.  Do I misunderstand?

**Comment by cdrini** [comment]
That is correct, but in order to update that symlink, we need to know that this repo's master is production ready.

**Comment by cdrini** [comment]
(Oh, just for clarity, note it's not just a symlink; the symlink points to a github submodule in `/vendors`)

**Comment by cclauss** [comment]
What are the git commands to update that submodule?

**Comment by cdrini** [comment]
Looks like you `cd` into the submodule directory run a `git pull` and then go back to the main and you should be able to commit the change (src: https://stackoverflow.com/a/8191413/2317712 ).

**Comment by hornc** [comment]
Closing this issue as resolved since I have reset master to a good condition, which has backed out #5, and made it so all the existing tests which still have value are running (and passing!) on CI under Python 2.7 (#28) I have also tagged the production 'release' that is currently being used by Open Library https://github.com/internetarchive/infogami/releases/tag/OL_201908 -- this is something we should continue to do as we update and commit the infogami submodule in https://github.com/internetarchive/openlibrary

The changes in #5 were good, and should be replayed, but perhaps in separate chunks, focusing on one specific fix type per PR, for ease of review and general management. We are properly committing to this course, and want to proceed with moving towards python3 compatibility on our master, production ready, branch. 

[I believe this comment represents the consensus agreed upon above, and in our other discussions. Repeating things as a summary and confirmation of our direction from now] 
Thank you everyone above for your comments and feedback! @cclauss @cdrini @mekarpeles
